stdlib/array.ml

```diff
@@ -292,3 +292,72 @@ let stable_sort cmp a =
 
 
 let fast_sort = stable_sort
+
+(* Array.filter and Bytes.filter are almost the same code:
+   if one needs to be modified, the other will probably need
+   it too *)
+
+let filter p arg =
+  let size_arg = length arg in
+  if size_arg = 0 then [| |] else
+    (* create the bit vector *)
+    let bv =
+      Bytes.create
+        (if size_arg land 7 == 0 then size_arg lsr 3
+         else (size_arg lsr 3) + 1) in
+    (* count elements e such that p(e) *)
+    let size_res = ref 0 in
+    (* fill bv, one byte at a time *)
+    for i = 0 to (size_arg lsr 3) - 1 do
+      let i8 = i lsl 3 in
+      (* p(arg.(8*i + k)) is (bv.[i] land (1 lsl k)) <> 0 *)
+      let c =
+        0 lor
+        (if p (unsafe_get arg i8)         then   1 else 0) lor
+        (if p (unsafe_get arg (i8 lor 1)) then   2 else 0) lor
+        (if p (unsafe_get arg (i8 lor 2)) then   4 else 0) lor
+        (if p (unsafe_get arg (i8 lor 3)) then   8 else 0) lor
+        (if p (unsafe_get arg (i8 lor 4)) then  16 else 0) lor
+        (if p (unsafe_get arg (i8 lor 5)) then  32 else 0) lor
+        (if p (unsafe_get arg (i8 lor 6)) then  64 else 0) lor
+        (if p (unsafe_get arg (i8 lor 7)) then 128 else 0) in
```

Comment:
Is that sequence better than a for loop ?
```OCaml
let accum = ref 0 in
for i = 0 to 7 do
  if p (unsafe_get arg (i8 lor i)) then accum := !accum lor (1 lsl i)    
done
```
The reduction in code size might be beneficial, and the single call to `p` might make the filter function more amenable to inlining. When the predicate is not inlined, the benefit from unrolling that loop might be lost in the noise.

Reply:
On my benchmarks (without flambda and where `p` tests `n mod k`), a `for` loop is 7% slower than my original code (on average).

I understand your point about inlining, but since Array.filter is meant to be a library function, inlining of `p` will happen only if `Array.filter` gets inlined itself, isn't it? Please confirm that it may happen and I'll make the change.

Thanks,